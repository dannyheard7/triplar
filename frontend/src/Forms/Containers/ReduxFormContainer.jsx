import React from "react";
import {inputParsers} from "../utils/inputParsers.js";
import {connect} from "react-redux";

class ReduxFormContainer extends React.Component {
    constructor(props) {
        super(props);

        this.handleSubmit = this.handleSubmit.bind(this);
    }

    handleSubmit(event) {
        event.preventDefault();
        const data = this.parseFormData(event);

        this.props.dispatch(this.props.action(data));
    }

    parseFormData(event) {
        const form = event.target;
        const data = new FormData(event.target);

        for (let name of data.keys()) {
            const input = form.elements[name];
            const parserName = input.dataset.parse;

            if (parserName) {
                const parser = inputParsers[parserName];
                const parsedValue = parser(data.get(name));
                data.set(name, parsedValue);
            }
        }

        let variables = Object();
        for(let pair of data.entries()) {
           variables[pair[0]] = pair[1];
        }

        return variables;
    }

    render() {
        const props = {
            onSubmit: this.handleSubmit,
            fieldErrors: this.props.fieldErrors,
            nonFieldErrors: this.props.nonFieldErrors

        };

        return React.cloneElement(this.props.children, {...props});
    }
}

ReduxFormContainer.defaultProps = {
    fieldErrors: [],
    nonFieldErrors: []
}

export default connect()(ReduxFormContainer)