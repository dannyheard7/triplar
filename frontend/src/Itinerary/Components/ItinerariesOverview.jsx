import React from "react";
import {connect} from "react-redux";
import {getTripItineraries} from "../utils/actions";
import {Link} from "react-router-dom";
import Moment from "moment";
import {faPlus} from "@fortawesome/free-solid-svg-icons";
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";

import 'Itinerary/styles/itineraries.css';
import LocationAddContainer from "./LocationAdd";


export class ItinerariesOverview extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            showLocationAdd: false
        };

        this.onClick = this.onClick.bind(this);
    }

    onClick() {
        this.setState({showLocationAdd: !this.state.showLocationAdd});
    }

    componentDidMount() {
        this.props.dispatch(getTripItineraries(this.props.match.params.id));
    }

    render() {
        const trip = this.props.trip;
        const locations = this.props.locations;

        return (
            <div className="itineraries-overview">
                <h2>Locations</h2>
                <div className="itineraries-list">
                    {locations.map((location) =>
                        <div className="card" key={location.id}>
                            <Link to={`/trips/${trip.id}/itinerary/${location.id}`}>
                                <h3 className="card-title">{location.city.name}{location.city.country && ", " + location.city.country.name}</h3>
                            </Link>
                            <p className="card-text">
                                {Moment(location.startDate).format('Do MMMM')} - {Moment(location.endDate).format('Do MMMM')}
                            </p>
                        </div>
                    )}
                    <div className="card">
                        {this.state.showLocationAdd ?
                            <LocationAddContainer trip={trip}/> :
                            <button className="btn btn-primary" onClick={this.onClick}><FontAwesomeIcon icon={faPlus}/></button>
                        }
                    </div>
                </div>
            </div>
        )
    }
}

ItinerariesOverview.defaultProps = {
    trip: {},
    locations: []
};

function mapStateToProps(state, ownProps) {
    return {
        trip: state.trips.trips.find(x => x.id === ownProps.match.params.id),
        locations: state.itineraries.locationItineraries.filter(x => x.trip.id === ownProps.match.params.id)
    };
}

export default connect(mapStateToProps)(ItinerariesOverview)
