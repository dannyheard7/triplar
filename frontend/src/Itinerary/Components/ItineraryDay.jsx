import React from "react";
import Moment from 'moment';

export default function ItineraryDay({day}) {
    day = Moment(day);

    return (
        <p>
            {day.format("dddd Do MMMM")}
        </p>
    );
}
