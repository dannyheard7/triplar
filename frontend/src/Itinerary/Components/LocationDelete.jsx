import React from "react";

import Modal from 'react-responsive-modal';
import {connect} from "react-redux";
import {deleteTripLocation} from "../utils/actions";
import LoadingIndicator from "../../App/Components/LoadingIndicator";

export class LocationDelete extends React.Component {
    constructor(props) {
        super(props);

        this.onDelete = this.onDelete.bind(this);
        this.closeModal = this.closeModal.bind(this);
    }

    closeModal() {
        // Should do this or rewrite path?
        this.props.history.goBack();
    }

    onDelete(e) {
        this.props.dispatch(deleteTripLocation(this.props.itinerary.id, this.props.match.params.tripId));
    };

    render() {
        if(this.props.itinerary) {
            return (
                <Modal open={true} onClose={this.closeModal} contentLabel={this.props.itinerary.city.name} >
                    <h2>Remove {this.props.itinerary.city.name}?</h2>

                    <p>Are you sure you want to remove the location '{this.props.itinerary.city.name}'?</p>

                    <button onClick={this.closeModal}>Close</button>
                    <button className="btn-danger" onClick={this.onDelete}>Remove</button>
                </Modal>
            );
        } else {
            return <LoadingIndicator/>
        }
    }
}

function mapStateToProps(state, ownProps) {
  return {
      itinerary: state.itineraries.locationItineraries.find(x => x.id === ownProps.match.params.itineraryId),
  };
}

export default connect(mapStateToProps)(LocationDelete)