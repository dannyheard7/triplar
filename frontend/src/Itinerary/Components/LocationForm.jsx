import React from "react";
import NonFieldErrors from "../../Forms/Components/NonFieldErrors";
import CitySearchFieldContainer from "../../Forms/Containers/CitySearchFieldContainer";
import DateFieldGroup from "../../Forms/Components/DateFieldGroup";
import moment from "moment";

moment.locale('en');

export default class LocationForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            arrivalDate: props.location.arrivalDate ? moment(props.location.arrivalDate) : moment(),
            departureDate: props.location.departureDate ? moment(props.location.departureDate) : moment()
        };

        this.onDateChange = this.onDateChange.bind(this);
        this.calculateEndDate = this.calculateEndDate.bind(this);
    }

    componentDidUpdate(prevProps, prevState) {
        if(prevProps.location.arrivalDate !== this.props.location.arrivalDate || prevProps.location.departureDate !== this.props.location.departureDate) {
            this.setState( {
                arrivalDate: moment(this.props.location.arrivalDate),
                departureDate: moment(this.props.location.departureDate)
            });
        }
    }

    onDateChange(value, date) {
        this.setState({
            [value]: moment(date, "DD/MM/YYYY")
        });
    }

    calculateEndDate(startDate) {
        const location = this.props.location;

        if(this.state.departureDate && this.state.departureDate.format("YYYY-MM-DD") !== location.departureDate) {
            return this.state.departureDate;
        } else {
            return startDate.clone().add(1, 'days');
        }
    }

    render() {
        const fieldErrors = this.props.fieldErrors;
        const itinerary = (this.props.location) ? this.props.location : {};
        const location = itinerary.city ? itinerary.city.name_std + ", " + itinerary.city.country : "";

        const arrivalDate = this.state.arrivalDate;
        const departureDate = this.calculateEndDate(this.state.arrivalDate);

        return (
            <form onSubmit={this.props.onSubmit}>
                <NonFieldErrors errors={this.props.nonFieldErrors}/>
                <CitySearchFieldContainer errors={fieldErrors.locations} value={location} name="city"/>
                <DateFieldGroup errors={fieldErrors.arrivalDate} label="Arrival" name="arrivalDate"
                                value={arrivalDate.format("DD/MM/YYYY")} required={true}
                                onChange={(date) => this.onDateChange("arrivalDate", date)}/>
                <DateFieldGroup errors={fieldErrors.departureDate} label="Departure" name="departureDate"
                                value={departureDate.format("DD/MM/YYYY")} required={true}
                                onChange={(date) => this.onDateChange("departureDate", date)}/>
                <button type="submit" className="btn btn-primary">{this.props.submitLabel}</button>
            </form>
        );
    }
}
