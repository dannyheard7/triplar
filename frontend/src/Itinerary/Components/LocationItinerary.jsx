import React from "react";
import Moment from 'moment';
import {daysBetweenDates} from 'Trips/utils/utils.js';
import PlacesSearchContainer from "Places/Containers/PlacesSearchContainer";
import ItineraryDayContainer from "Itinerary/Containers/ItineraryDayContainer";
import {connect} from "react-redux";
import {getItineraryDayDetail} from "../utils/actions";
import MarkerMap from "../../Maps/Components/MarkerMap";


export class LocationItinerary extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            showMap: false,
            selectedDate: props.itinerary.startDate,
        };

        this.onClick = this.onClick.bind(this);
        this.onChangeSelectedDay = this.onChangeSelectedDay.bind(this);
    }

    componentDidMount() {
        this.props.dispatch(getItineraryDayDetail(this.props.itinerary.id, this.state.selectedDate));
    }

    onClick() {
        this.setState({showMap: !this.state.showMap});
    }

    onChangeSelectedDay(event) {
        this.setState({selectedDate: event.target.value});
        this.componentDidMount();
    }

    render () {
        const itinerary = this.props.itinerary;
        const days = daysBetweenDates(itinerary.startDate, itinerary.endDate);

        return (
            <div className="destination-itinerary">
                <div className="destination-plan">
                    <h3 onClick={this.onClick}>{itinerary.city.name}</h3>
                    <p>{Moment(itinerary.startDate).format('dddd Do MMMM') } - {Moment(itinerary.endDate).format('dddd Do MMMM') }
                        <select onChange={this.onChangeSelectedDay}>
                            {days.map((day) => <option key={day} value={day.format("YYYY-MM-DD")}>{day.format('dddd Do MMMM')}</option>)}
                        </select>
                    </p>
                    <ItineraryDayContainer day={this.state.selectedDate} itinerary={itinerary} key={this.state.selectedDate}
                        path={`/trips/${this.props.tripId}`} />
                </div>
                {itinerary.city.location && <PlacesSearchContainer city={itinerary.city} path={`/trips/${this.props.tripId}`}/>}
            </div>
        )
    }
}

LocationItinerary.defaultProps = {
    itinerary: {
        city: {
            location: {
                lat: 0,
                lng: 0
            }
        }
    }
};

function mapStateToProps(state, ownProps) {
  return {
      itinerary: state.itineraries.locationItineraries.find(x => x.id === ownProps.location.id)
  };
}

export default connect(mapStateToProps)(LocationItinerary)
