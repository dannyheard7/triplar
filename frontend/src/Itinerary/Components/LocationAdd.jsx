import React from "react";
import LocationForm from "./LocationForm";
import ReduxFormContainer from "../../Forms/Containers/ReduxFormContainer";
import {addTripLocation} from "../utils/actions";
import {connect} from "react-redux";

export class LocationAdd extends React.Component {
    curryAction = (action, id) => (object) => action(id, object);

    render () {
        const errors = this.props.errors ? this.props.errors : [];

        const location = {arrivalDate: this.props.trip.startDate, department: this.props.trip.endDate};

        return (
            <ReduxFormContainer action={this.curryAction(addTripLocation, this.props.trip.id)} fieldErrors={errors} >
                <LocationForm submitLabel="Add" location={location}/>
            </ReduxFormContainer>
        )
    }
}

const mapStateToProps = (state) => ({
    errors: state.errors.ADD_TRIP_LOCATION
});
export default connect(mapStateToProps)(LocationAdd)