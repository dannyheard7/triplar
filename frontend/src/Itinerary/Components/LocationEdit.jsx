import React from "react";
import LocationForm from "./LocationForm";
import ReduxFormContainer from "../../Forms/Containers/ReduxFormContainer";
import {addTripLocation} from "../utils/actions";
import {connect} from "react-redux";

export class LocationAdd extends React.Component {
    curryAction = (action, id) => (object) => action(id, object);

    render () {
        return (
            <ReduxFormContainer action={this.curryAction(addTripLocation, this.props.trip.id)} errors={this.props.errors} >
                <LocationForm submitLabel="Add" location={this.props.trip}/>
            </ReduxFormContainer>
        )
    }
}

const mapStateToProps = (state) => {return {errors: state.itineraries.errors}};
export default connect(mapStateToProps)(LocationAdd)