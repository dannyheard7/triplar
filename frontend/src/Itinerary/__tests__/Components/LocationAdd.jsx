import React from "react";
import {shallow} from "enzyme";

import ShallowRenderer from 'react-test-renderer/shallow';
import {LocationAdd} from "../../Components/LocationAdd";
import ReduxFormContainer from "../../../Forms/Containers/ReduxFormContainer";

const renderer = new ShallowRenderer();
const faker = require('faker');

jest.mock('../../utils/itinerary.api.js');

const api = require('../../utils/itinerary.api.js');

describe('<LocationAdd />', () => {
    const data = {tripLocation: {'id': faker.random.number(),
                'city': {'name_std': faker.address.city(), 'country': faker.address.country()}}};

    test('renders correctly', () => {
        const result = renderer.render(<LocationAdd trip={{id: 1, startDate: '2018-08-14', endDate: '2018-08-18'}} />);
        expect(result).toMatchSnapshot();
    });

    // TODO: Move to redux
    test.skip('calls onSuccess after receiving success from <FormContainer />', (function () {
        const spy = jest.spyOn(LocationAddContainer.prototype, 'onSuccess');
        const stub = jest.fn();
        const container = shallow(<LocationAdd trip={{id: faker.random.number()}} onSuccess={stub} />);
        container.setState({showLocationAdd: true});
        container.find(ReduxFormContainer).prop('onSuccess')(data);

        expect(spy).toBeCalledWith(data);
        expect(stub).toBeCalledWith(data.tripLocation)
    }));

    // TODO: Move to redux
    test.skip('addTripLocation called with correct arguments', () => {
        const spy = jest.spyOn(api.default, 'addTripLocation');
        const trip = {id: faker.random.number()};

        const container = shallow(<LocationAdd trip={trip} />);
        container.instance().apiFunc(data);

        expect(spy).toBeCalledWith(trip.id, data);

        spy.mockRestore();
    });




        // TODO
    test.skip('does not render <ReduxFormContainer /> after a trip has been created', () => {
        expect(container.find(ReduxFormContainer).length).toEqual(1);
        container.find(ReduxFormContainer).prop('onSuccess')(data);
        container.update();

        expect(container.state('showLocationAdd')).toEqual(false);
        expect(container.find(ReduxFormContainer).length).toEqual(0);
    });

});