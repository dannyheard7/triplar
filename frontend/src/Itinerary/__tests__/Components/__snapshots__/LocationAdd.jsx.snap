// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LocationAdd /> renders correctly 1`] = `
<Connect(ReduxFormContainer)
  action={[Function]}
  fieldErrors={Array []}
>
  <LocationForm
    location={
      Object {
        "id": 1,
      }
    }
    submitLabel="Add"
  />
</Connect(ReduxFormContainer)>
`;
